transaction/std/src/memo/gift_code_sender.rs

```diff
@@ -27,23 +29,62 @@ impl RegisteredMemoType for GiftCodeSenderMemo {
 }
 
 impl GiftCodeSenderMemo {
-    /// The length of the custom memo data.
+    /// Number of bytes in the memo
     pub const MEMO_DATA_LEN: usize = 64;
 
-    /// Create a new gift code memo
-    pub fn new(note_data: &str) -> Result<Self, MemoError> {
-        GiftCodeSenderMemo::try_from(note_data)
+    /// Number of bytes used to represent the fee paid for gift code redemption
+    pub const FEE_DATA_LEN: usize = 7;
+
+    /// Number of bytes used to represent the utf-8 note
+    pub const NOTE_DATA_LEN: usize = 57;
+
+    /// The max fee (i.e. the maximum value a 56 bit number can contain)
+    pub const MAX_FEE: u64 = u64::MAX >> 8;
+
+    /// Create a new gift code sender memo
+    pub fn new(fee: u64, note: &str) -> Result<Self, MemoError> {
+        // Check if the fee we're setting is greater than the max fee
+        if fee > Self::MAX_FEE {
+            return Err(MemoError::MaxFeeExceeded(Self::MAX_FEE, fee));
+        }
+        // Check if note is of valid length
+        if note.len() > Self::NOTE_DATA_LEN {
+            return Err(MemoError::BadLength(note.len()));
+        }
+
+        let mut memo_data = [0u8; Self::MEMO_DATA_LEN];
+
+        // Put fee into memo
+        memo_data[0..Self::FEE_DATA_LEN].copy_from_slice(&fee.to_be_bytes()[1..]);
+
+        // Put note into memo
+        memo_data[Self::FEE_DATA_LEN..(Self::FEE_DATA_LEN + note.len())]
```

Comment:
Is the upper bound necessary?
```suggestion
        memo_data[Self::FEE_DATA_LEN..]
```

Reply:
Lower bound, and yeah because otherwise we're trying to copy a note of varied length into exactly 57 bytes